brave/brave-core
# Implement text embedding processing for ad matching MVP

Resolves https://github.com/brave/brave-browser/issues/23424

This PR adds functionality to compute text embeddings (numerical representations) in the client. Specifically, embeddings are computed for visited webpages based on their og tags, if available. A history of k recent embeddings are stored locally (k can be controlled by a griffin parameter). In a subsequent PR, these embeddings will be used for client-side ad targeting.  

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues/23424) for my issue

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

To test the embedding functionality, consider the following:
- build browser locally
- run browser with `npm start -- --enable-logging=stderr --vmodule=text_embedding_processor=9,embedding_processing=9,text_embedding_html_events=9 --enable-features=TextEmbedding` for relevant logging and dev component resources
 - opt-in to ads
 - restart browser
 - go to webpages with og:title tags (e.g. many wikipedia pages have these). If the title has in-vocabulary words, an embedding vector will be computed and logged
 - open sqlite terminal in `Brave-Browser-Development/Default/ads_service` via `sqlite3 database.sqlite`
 - check that embeddings being stored in table `text_embedding_html_events`
 - check that no more than 10 (configurable parameter) embeddings are stored in `text_embedding_html_events`. additionally check that embedding row is updated (not re-inserted) if duplicate in table (can identify embedding by hashed_key column)
